backend/src/apis/snapshot_data.rs

```diff
@@ -0,0 +1,286 @@
+use num_bigint::BigInt;
+use std::collections::HashMap;
+
+use halo2_proofs::{
+    halo2curves::bn256::{Bn256, Fr as Fp, G1Affine},
+    plonk::{keygen_pk, keygen_vk, ProvingKey, VerifyingKey},
+    poly::kzg::commitment::ParamsKZG,
+};
+use snark_verifier_sdk::CircuitExt;
+
+use summa_solvency::{
+    circuits::{
+        merkle_sum_tree::MstInclusionCircuit,
+        solvency::SolvencyCircuit,
+        utils::{full_prover, generate_setup_params},
+    },
+    merkle_sum_tree::utils::big_int_to_fp,
+    merkle_sum_tree::MerkleSumTree,
+};
+
+use crate::apis::csv_parser::parse_csv_to_assets;
+
+struct MstParamsAndKeys<const LEVELS: usize, const L: usize, const N_ASSETS: usize> {
+    params: ParamsKZG<Bn256>,
+    pk: ProvingKey<G1Affine>,
+    vk: VerifyingKey<G1Affine>,
+}
+
+struct SnapshotData<
+    const LEVELS: usize,
+    const L: usize,
+    const N_ASSETS: usize,
+    const N_BYTES: usize,
+    const K: u32,
+> {
+    exchange_id: String,
+    mst: MerkleSumTree<N_ASSETS>,
+    assets: Vec<Asset>,
+    mst_params_and_keys: MstParamsAndKeys<LEVELS, L, N_ASSETS>,
+    proofs_of_inclusions: HashMap<u64, UserProof>,
+    proof_of_solvency: Option<SolvencyProof<N_ASSETS>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Asset {
+    pub name: String,
+    pub pubkeys: Vec<String>,
+    pub balances: Vec<BigInt>,
+    pub sum_balances: Fp,
+    pub signatures: Vec<String>,
```

Comment:
Currently, the contract will take the exchange addresses and scan all of them to calculate the total sum of a particular asset, but the signature check is a separate process. Since the contract takes the CEX addresses as an independent argument, it simply checks each address' signature while iterating over the addresses. Do you think that it makes sense to attach signatures to an asset here? This way multiple assets could have the same set of signatures.

Reply:
I got your point! You are right, we need to separate the signatures from here to avoid storing duplicate signatures.

Reply:
Updated at [771ad22](https://github.com/summa-dev/summa-solvency/pull/88/commits/771ad223da27f5dd9d15e5e3d63d644aae4289f9)